### PR TITLE
[Eager, Performance optimization] math op sink to Cpp level ( + and - operator as an example )

### PR DESCRIPTION
**Pull request by veyron95**
### PR types

Performance optimization
### PR changes

APIs
### Describe

This PR support **+** and **-** operator expose under Cpp level directly.

### CODE REVIEW

File: paddle/fluid/pybind/eager_math_op_patch.cc [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+      ret = CallScalarFuction(self_tensor, other, "add");
+    }
+    return ToPyObject(ret);
+  }
+
+  // 2. create or get tensor for other_obj
+  paddle::experimental::Tensor other_tensor;
+  if (!PyCheckTensor(other_obj)) {
+    paddle::experimental::Scalar value = CastPyArg2Scalar(other_obj, "full", 0);
+    {
```
Comment (wanghuancoder) [inline paddle/fluid/pybind/eager_math_op_patch.cc]: full改成__add__，这个适用于报错给用户的。因为是在__add__运算符里，所以应该报__add__

File: paddle/fluid/pybind/eager_math_op_patch.cc [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+    // dtype
+    if (_complex_dtypes.find(lhs_dtype) != _complex_dtypes.end() ||
+        _complex_dtypes.find(rhs_dtype) != _complex_dtypes.end()) {
+      phi::DataType promote_dtype =
+          framework::TransToPhiDataType(framework::PromoteTypesIfComplexExists(
+              framework::TransToProtoVarType(lhs_dtype),
+              framework::TransToProtoVarType(rhs_dtype)));
+      if (lhs_dtype != promote_dtype) {
+        // cast
+        self_tensor = cast_ad_func(self_tensor, promote_dtype);
```
Comment (wanghuancoder) [inline paddle/fluid/pybind/eager_math_op_patch.cc:216]: cast也需要释放gil

File: paddle/fluid/pybind/eager_math_op_patch.cc (hunk 1 above)
Comment (veyron95) [reply paddle/fluid/pybind/eager_math_op_patch.cc]: 已修改，感谢。

File: paddle/fluid/pybind/eager_math_op_patch.cc (hunk 2 above)
Comment (veyron95) [reply paddle/fluid/pybind/eager_math_op_patch.cc:216]: 已补充，感谢。

**Review by wanghuancoder — APPROVED** [review]
LGTM

**Pull request merged**